scheduler/context.go

```diff
@@ -124,3 +129,121 @@ func (e *EvalContext) ProposedAllocs(nodeID string) ([]*structs.Allocation, erro
 	}
 	return proposed, nil
 }
+
+func (e *EvalContext) Eligibility() *EvalEligibility {
+	if e.eligibility == nil {
+		e.eligibility = NewEvalEligibility()
+	}
+
+	return e.eligibility
+}
+
+type ComputedClassEligibility byte
+
+const (
+	// The EvalComputedClass enums denote the eligibility of the computed class
+	// for the evaluation.
+	EvalComputedClassUnknown ComputedClassEligibility = iota
+	EvalComputedClassIneligible
+	EvalComputedClassEligible
+	EvalComputedClassEscaped
+)
+
+// EvalEligibility tracks eligibility of nodes by computed node class over the
+// course of an evaluation.
+type EvalEligibility struct {
+	// Job tracks the eligibility at the job level per computed node class.
+	Job map[uint64]ComputedClassEligibility
+
+	// JobEscapedConstraints tracks escaped constraints at the job level.
+	JobEscapedConstraints []*structs.Constraint
+
+	// TaskGroups tracks the eligibility at the task group level per computed
+	// node class.
+	TaskGroups map[string]map[uint64]ComputedClassEligibility
+
+	// TgEscapedConstraints is a map of task groups to a set of constraints that
+	// have escaped.
+	TgEscapedConstraints map[string][]*structs.Constraint
+}
+
+// NewEvalEligibility returns an eligibility tracker for the context of an evaluation.
+func NewEvalEligibility() *EvalEligibility {
+	return &EvalEligibility{
+		Job:                  make(map[uint64]ComputedClassEligibility),
+		TaskGroups:           make(map[string]map[uint64]ComputedClassEligibility),
+		TgEscapedConstraints: make(map[string][]*structs.Constraint),
+	}
+}
+
+// SetJob takes the job being evaluated and calculates the escaped constraints
+// at the job and task group level.
+func (e *EvalEligibility) SetJob(job *structs.Job) {
+	// Determine the escaped constraints for the job.
+	e.JobEscapedConstraints = structs.EscapedConstraints(job.Constraints)
+
+	// Determine the escaped constraints per task group.
+	for _, tg := range job.TaskGroups {
+		constraints := tg.Constraints
+		for _, task := range tg.Tasks {
+			constraints = append(constraints, task.Constraints...)
+		}
+
+		e.TgEscapedConstraints[tg.Name] = structs.EscapedConstraints(constraints)
+	}
+}
+
+// JobStatus returns the eligibility status of the job.
+func (e *EvalEligibility) JobStatus(class uint64) ComputedClassEligibility {
```

Comment:
Class value `0` should be special cased, e.g. nodes that do not yet have one computed (legacy migration).